src/datepicker/bs-daterangepicker.component.ts

```diff
@@ -251,12 +252,46 @@ export class BsDaterangepickerDirective
         dateCustomClasses: this.dateCustomClasses || this.bsConfig && this.bsConfig.dateCustomClasses,
         datesDisabled: this.datesDisabled || this.bsConfig && this.bsConfig.datesDisabled,
         datesEnabled: this.datesEnabled || this.bsConfig && this.bsConfig.datesEnabled,
-        ranges: this.bsConfig && this.bsConfig.ranges,
+        ranges: this.checkMaxRangeDate(this.bsConfig && this.bsConfig.ranges, this.maxDate || this.bsConfig && this.bsConfig.maxDate),
         maxDateRange: this.bsConfig && this.bsConfig.maxDateRange
       }
     );
   }
 
+  /**
+   * Checks max date and dates in ranges
+   * if one of dates in ranges is later then maxDate
+   * this date becomes equal to maxDate
+   */
+
+  checkMaxRangeDate(ranges?: BsCustomDates[], maxDate?: Date): BsCustomDates[] | void {
+    if (!ranges || !maxDate || (ranges && !ranges.length && !ranges[0].value)) return ranges;
```

Comment:
Please extract this method to helper file and cover with tests.
And don't nest ifs, prefer guard checks and early exits